include/insn_queue.h

```diff
@@ -0,0 +1,132 @@
+/*
+ *  libpulp - User-space Livepatching Library
+ *
+ *  Copyright (C) 2023 SUSE Software Solutions GmbH
+ *
+ *  This file is part of libpulp.
+ *
+ *  libpulp is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ *  libpulp is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with libpulp.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef INSNQ_H
+#define INSNQ_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+/** Define a 2Mb buffer for holding the instructions list.  */
+#define INSN_BUFFER_MAX (2 * 1024 * 1024)
+
+/** The ULP instruction queue.  This works as follows:
+ * 1- Libpulp write instructions that should be executed on the `ulp` tool
+ * side. The main reason why this exists is because some processes are launched
+ * with systemd memory protection mechanism, and a way to circunvent that is to
+ * ptrace from an external process.
+ *
+ * 2- the `ulp` tool interpret the instructions after libpulp finishes
+ * executing the function from its side (__ulp_apply_patch, for example).
+ */
+struct insn_queue
+{
+  /** Number of instructions in this queue.  */
+  int num_insns;
+
+  /** Size in bytes of content.  Must not be larger than INSN_BUFFER_MAX.  */
+  int size;
+
+  /** Buffer holding the instructions.  */
+  char buffer[INSN_BUFFER_MAX];
+};
```

Comment:
A new field named `version` was introduced to this structure. The ULP tool now checks this field to ensure that itself is not old enough, and skips the process with an error in case it is not compatible.